lib/Dialect/SV/Transforms/HWRaiseInOutPorts.cpp

```diff
@@ -0,0 +1,195 @@
+  for (auto *user : body->getArgument(port.argNum).getUsers()) {
+    if (auto read = dyn_cast<sv::ReadInOutOp>(user))
+      readers.push_back(read);
+    else if (auto write = dyn_cast<sv::AssignOp>(user))
+      writers.push_back(write);
+    else
+      user->emitWarning() << "Use of inout port " << port.name
+                          << " is not supported";
+  }
+
+  if (writers.size() > 1)
+    converter.getModule()->emitWarning()
+        << "Multiple writers of inout port " << port.name
+        << " detected. Will only create an output for the first write";
+}
+
+void HWInOutPortConversion::buildInputSignals() {
+  if (hasReaders()) {
+    Value readValue =
+        converter.createNewInput(origPort, "_rd", origPort.type, readPort);
+
+    // Replace all sv::ReadInOutOp's with the new input.
+    Value origInput = body->getArgument(origPort.argNum);
+    for (auto *user : llvm::make_early_inc_range(origInput.getUsers())) {
+      sv::ReadInOutOp read = dyn_cast<sv::ReadInOutOp>(user);
+      if (!read)
+        continue;
+
+      read.replaceAllUsesWith(readValue);
+      read.erase();
+    }
+  }
+
+  if (hasWriters()) {
+    sv::AssignOp write = writers.front();
+    converter.createNewOutput(origPort, "_wr", origPort.type, write.getSrc(),
+                              writePort);
+    write.erase();
+  }
+}
+
+void HWInOutPortConversion::buildOutputSignals() {
+  // TODO: could support hw.inout outputs (always create read/write ports) -
+  // don't need it for now, though.
+  assert(false && "hw.inout outputs not yet supported");
+}
+
+void HWInOutPortConversion::mapInputSignals(OpBuilder &b, Operation *inst,
+                                            Value instValue,
+                                            SmallVectorImpl<Value> &newOperands,
+                                            ArrayRef<Backedge> newResults) {
+
+  if (hasReaders()) {
+    // Create a read_inout op at the instantiation point. This effectively
+    // pushes the read_inout op from the module to the instantiation site.
+    newOperands[readPort.argNum] =
+        b.create<ReadInOutOp>(inst->getLoc(), instValue).getResult();
+  }
+
+  if (hasWriters()) {
+    // Create a sv::AssignOp at the instantiation point. This effectively
+    // pushes the write op from the module to the instantiation site.
+    Value writeFromInsideMod = newResults[writePort.argNum];
+    b.create<sv::AssignOp>(inst->getLoc(), instValue, writeFromInsideMod);
+  }
+}
+
+void HWInOutPortConversion::mapOutputSignals(
+    OpBuilder &b, Operation *inst, Value instValue,
+    SmallVectorImpl<Value> &newOperands, ArrayRef<Backedge> newResults) {
+  llvm_unreachable("hw.inout outputs not yet supported");
+}
+
+class HWInoutPortConversionBuilder : public PortConversionBuilder {
+public:
+  using PortConversionBuilder::PortConversionBuilder;
+  FailureOr<std::unique_ptr<PortConversion>> build(hw::PortInfo port) override {
+    if (port.direction == hw::PortDirection::INOUT)
```

Comment:
Since we don't (yet) support output inouts, we should either ignore them or issue an error. Unsupported IR should never assert or segfault, as I think this results in.

Reply:
Won't segfault, just doesn't do that much - e.g.

```mlir
hw.module @write(%a: !hw.inout<i42>) {
  %0 = hw.constant 0 : i42
  sv.assign %a, %0 : i42
}

hw.module @outputInout() -> (out : !hw.inout<i42>) {
  %0 = sv.wire : !hw.inout<i42>
  hw.output %0 : !hw.inout<i42>
}

hw.module @outputInoutDriver() {
  %0 = hw.instance "outputInout" @outputInout() -> (out : !hw.inout<i42>)
  hw.instance "write" @write(a : %0 : !hw.inout<i42>) -> ()
}
```

results in

```mlir
hw.module @write() -> (a_wr: i42) {
    %c0_i42 = hw.constant 0 : i42
    hw.output %c0_i42 : i42
  }

hw.module @outputInout() -> (out: !hw.inout<i42>) {
    %0 = sv.wire : !hw.inout<i42>
    hw.output %0 : !hw.inout<i42>
  }
  hw.module @outputInoutDriver() {
    %outputInout.out = hw.instance "outputInout" @outputInout() -> (out: !hw.inout<i42>)
    sv.assign %outputInout.out, %write.a_wr : i42
    %write.a_wr = hw.instance "write" @write() -> (a_wr: i42)
    hw.output
  }
```

Reply:
Yep. I forgot that confusing behavior. I'm hoping that'll change with lenharth's moduletype.